micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jpa/HibernateQueryMetrics.java

```diff
@@ -0,0 +1,181 @@
+            MetricsEventHandler metricsEventHandler = new MetricsEventHandler(meterRegistry);
+            registry.appendListeners(EventType.POST_LOAD, metricsEventHandler);
+        }
+
+    }
+
+    class MetricsEventHandler implements PostLoadEventListener {
+
+        private final MeterRegistry meterRegistry;
+
+        MetricsEventHandler(MeterRegistry meterRegistry) {
+            this.meterRegistry = meterRegistry;
+        }
+
+        @Override
+        public void onPostLoad(PostLoadEvent event) {
+            registerQueryMetric(event.getSession().getFactory().getStatistics());
+        }
+
+        void registerQueryMetric(Statistics statistics) {
+            for (String query : statistics.getQueries()) {
+                QueryStatistics queryStatistics = statistics.getQueryStatistics(query);
+                if (Objects.nonNull(queryStatistics)) {
+                    String queryName = query.replace(" ", "_").replace(".", "_").toLowerCase();
+                    if (Objects.isNull(Search.in(meterRegistry).tags("query", queryName).functionCounter())) {
+
+                        FunctionCounter.builder("hibernate.query.cache.requests", queryStatistics, QueryStatistics::getCacheHitCount)
+                                .tags(tags)
+                                .tags("result", "hit", "query", queryName)
+                                .description("The number of query cache hits")
+                                .register(meterRegistry);
+
+                        FunctionCounter.builder("hibernate.query.cache.requests", queryStatistics, QueryStatistics::getCacheMissCount)
+                                .tags(tags)
+                                .tags("result", "miss", "query", queryName)
+                                .description("The number of query cache miss")
+                                .register(meterRegistry);
+
+                        FunctionCounter.builder("hibernate.query.cache.puts", queryStatistics, QueryStatistics::getCacheMissCount)
+                                .tags(tags)
+                                .tags("query", queryName)
+                                .description("The number of putting the query into cache")
+                                .register(meterRegistry);
+
+                        FunctionTimer.builder("hibernate.query.executions.total", queryStatistics, QueryStatistics::getExecutionCount, QueryStatistics::getExecutionTotalTime, TimeUnit.MILLISECONDS)
+                                .tags(tags)
+                                .tags("query", queryName)
+                                .description("Function tracked total number of query executions during time")
+                                .register(meterRegistry);
+
+                        TimeGauge.builder("hibernate.query.executions.avg", queryStatistics, TimeUnit.MILLISECONDS, QueryStatistics::getExecutionAvgTime)
```

Comment:
This is redundant because the `FunctionTimer` can already provide the average or it can be derived from the totalTime / count.

Reply:
Removed